src/JointPositionGoalController.cpp

```diff
@@ -0,0 +1,349 @@
+            max_joint_velocity_ = 10;  // Default to 10deg/s if not set
+        }
+        max_joint_velocity_rad_ = (max_joint_velocity_ * M_PI / 180);
+        ROS_INFO_STREAM("Maximum joint velocity: " << max_joint_velocity_ << "deg/s, " << max_joint_velocity_rad_ << "rad/s");
+
+        // Retrieve joint limits from parameter server
+        for (auto const &joint_name : joint_names_) {
+            joint_limits_interface::JointLimits limits;
+            if (!getJointLimits(joint_name, controller_nh, limits))
+                ROS_ERROR_STREAM("Cannot read joint limits for joint " << joint_name << " from param server");
+
+            joint_limits_.insert(std::make_pair(joint_name, limits));
+            ROS_INFO_STREAM("Joint Position Limits: " << joint_name << " has lower limit " << limits.min_position <<
+                            " and upper limit " <<
+                            limits.max_position);
+        }
+
+        // get a pointer to the position interface
+        hardware_interface::PositionJointInterface *position_hw = robot_hw->get<hardware_interface::PositionJointInterface>();
+        if (!position_hw) {
+            ROS_ERROR(
+                    "This controller requires a hardware interface of type hardware_interface::PositionJointInterface.");
+            return false;
+        }
+
+        position_hw->clearClaims();
+        const std::vector<std::string> &positionNames = position_hw->getNames();
+        // initialize command buffer for each joint we found on the HW
+        for (unsigned int i = 0; i < positionNames.size(); i++) {
+            if (use_joint_selection &&
+                std::find(joint_names_.begin(), joint_names_.end(), positionNames[i]) == joint_names_.end())
+                continue;
+
+            positionJointHandles_[positionNames[i]] = position_hw->getHandle(positionNames[i]);
+            latest_commands_[positionNames[i]] = 0.0;
+        }
+        q_move_time_ = 0;
+
+        auto position_hw_claims = position_hw->getClaims();
+        claimed_resources.insert(position_hw_claims.begin(), position_hw_claims.end());
+        position_hw->clearClaims();
+
+        // success
+        state_ = INITIALIZED;
+        ROS_INFO_STREAM(
+                "JointPositionGoalController ON with " << claimed_resources.size() << " claimed resources:" << std::endl
+                << position_hw_claims.size() << " position-controlled joints" << std::endl);
+        return true;
+    }
+
+    void JointPositionGoalController::starting(const ros::Time &time) {
+        last_update_ = time;
+    }
+
+    void JointPositionGoalController::update(const ros::Time &time, const ros::Duration &period) {
+        handler_->update();
+        lcm_->handleTimeout(0);
+
+        double dt = (time - last_update_).toSec();
+        int64_t utime = (int64_t) (time.toSec() * 1000000.);
+
+        size_t number_of_joint_interfaces = positionJointHandles_.size();
+
```

Comment:
Can we remove these diagnostics publishers (preferred) or share them with the other LCM2ROS controller.

They should be functionalised that than dropped into the middle of a long function